trezorlib/client.py

```diff
@@ -1100,6 +1101,172 @@ def self_test(self):
+                    source_account=op["source_account"],
+                    selling_asset=selling_asset,
+                    buying_asset=buying_asset,
+                    amount=op["amount"],
+                    price_n=op["price_n"],
+                    price_d=op["price_d"]
+                ))
+            # Set Options
+            if op["type"] == 5:
+                resp = self.call(proto.StellarSetOptionsOp(
+                    source_account=op["source_account"],
+                    inflation_destination_account=op["inflation_destination"],
+                    clear_flags=op["clear_flags"],
+                    set_flags=op["set_flags"],
+                    master_weight=op["master_weight"],
+                    low_threshold=op["low_threshold"],
+                    medium_threshold=op["medium_threshold"],
+                    high_threshold=op["high_threshold"],
+                    home_domain=op["home_domain"],
+                    signer_type=op["signer_type"],
+                    signer_key=op["signer_key"],
+                    signer_weight=op["signer_weight"],
+                ))
+            # Change Trust
+            if op["type"] == 6:
+                asset = types.StellarAssetType(type=op["asset"]["type"], code=op["asset"]["code"],
+                                               issuer=op["asset"]["issuer"])
+                resp = self.call(proto.StellarChangeTrustOp(
+                    source_account=op["source_account"],
+                    limit=op["limit"],
+                    asset=asset
+                ))
+            # Allow Trust
+            if op["type"] == 7:
+                resp = self.call(proto.StellarAllowTrustOp(
+                    source_account=op["source_account"],
+                    trusted_account=op["trusted_account"],
+                    asset_type=op["asset_type"],
+                    asset_code=op["asset_code"],
+                    is_authorized=op["is_authorized"]
+                ))
+            # Merge Account
+            if op["type"] == 8:
+                resp = self.call(proto.StellarAccountMergeOp(
+                    source_account=op["source_account"],
+                    destination_account=op["destination_account"]
+                ))
+            # Manage data
+            if op["type"] == 10:
+                resp = self.call(proto.StellarManageDataOp(
+                    source_account=op["source_account"],
+                    key=op["key"],
+                    value=op["value"]
+                ))
+            # Merge Account
+            if op["type"] == 11:
+                resp = self.call(proto.StellarBumpSequenceOp(
+                    source_account=op["source_account"],
+                    bump_to=op["bump_to"]
+                ))
+
+            # Exit if the response was a StellarSignedTx
+            if resp.__class__.__name__ == "StellarSignedTx":
+                return resp
```

Comment:
this seems that you simply drop every `resp` that isn't a `StellarSignedTx`? what are these?

Reply:
The signing loop for Stellar looks like this:

1. Start with a Transaction that contains a Header and N Operations
2. Send the Header to the device
3. Receive a `StellarTxOpRequest` message
4. Send Operation 1 (protobuf message varies depending on operation type)
5. Receive `StellarTxOpRequest` (device expects another operation) OR `StellarSignedTx` (that was the last Operation in the Transaction)
a. If a `StellarSignedTx` was received, return it since this is what we're after
b. Otherwise, go to 4 and send Operation 2

Since the python code and the device should both agree on the number of operations I could remove the check for whether a `StellarSignedTx` came back and instead move this outside the loop and return it after all operations were sent, if you think that would read better?